README.md
# EHDEV-CONFIGER-SPA
---

[![Build Status](https://travis-ci.org/EHDFE/ehdev-configer-spa.svg?branch=master)](https://travis-ci.org/EHDFE/ehdev-configer-spa)
[![npm](https://img.shields.io/npm/dm/ehdev-configer-spa.svg)]()
[![npm](https://img.shields.io/npm/v/ehdev-configer-spa.svg)]()
[![node](https://img.shields.io/node/v/ehdev-configer-spa.svg)]()
[![GitHub tag](https://img.shields.io/github/tag/ehdfe/ehdev-configer-spa.svg)]()
[![David](https://img.shields.io/david/EHDFE/ehdev-configer-spa.svg)]()
[![David](https://img.shields.io/david/dev/EHDFE/ehdev-configer-spa.svg)]()


## 配置说明

| 配置项 | 默认值 | 说明 |
|---|---| ---|
| buildPath | dist | 输出目录 |
| enableHotModuleReplacement | true | 启用[热更新](https://webpack.js.org/guides/hot-module-replacement) |
|framework|raect|依赖框架，目前只对 `react` 有做优化，包括引入 `react-hot-loader`|
|htmlWebpackPlugin|`{inject: true, chunksSortMode: 'auto', cache: true, showErrors: true}`|`htmlWebpackPlugin` 插件配置, 参考 [https://github.com/jantimon/html-webpack-plugin#configuration](https://github.com/jantimon/html-webpack-plugin#configuration)|
| browserSupports | last 2 version | 浏览器支持配置，影响 `babel` 和 `autoprefixer`, 配置参考：[https://github.com/ai/browserslist](https://github.com/ai/browserslist) |
| dll | `{ enable: false, enclude: [] }` | 是否启用 dll，enclude 提供打入 dll 包的模块 |
| providePluginConfig | `{}` | 主要用来支持 jQuery 依赖全局挂载的老模块， 参考 [https://webpack.js.org/plugins/provide-plugin/]
(https://webpack.js.org/plugins/provide-plugin/) |
| babelUseBuiltIns | true | [babel-preset-env#usebuiltins](http://babeljs.io/docs/plugins/preset-env/#usebuiltins) 配置 |
| https | false | 开发环境的 https 支持 |
| publicPath | `../` | `webpackConfig.output.publicPath`, 只在构建时生效 |
| contentBase | undefined | 配置 devServer 的 [contentBase](https://webpack.js.org/configuration/dev-server/#devserver-contentbase)，默认包含当前项目的输出目录，不需要配置 |

## 使用说明

### svg 用法

引用 svg 路径后加上 `?reactComponnet` ，svg 会被转换成 react component
否则 svg 会被当成普通的资源文件，使用 file-loader 加载

```js
import Foo from './foo.svg?reactComponnet';
<Foo />
```


HISTORY.md
# HISTORY

## v0.0.10
---
FEATURE:
  - add `supportIE8`
  
CHANGE:
  - remove `uglifyPlugin` from `prod.conf.js`

## v0.0.9
---
FEATURE:
  - add `publicPath` in projectConfig

## v0.0.8
---
FIX:
- add `CONFIGER_FOLDER_PATH` to resolve.modules which cause missing package issue

FEATURES:
- supports `svg to react component` feature.


test/lib.test.js
const path = require('path');
const { expect } = require('chai');

const { getFilesByExtName } = require('../src/lib');

describe('lib tests', () => {
  const srcDir = process.cwd();

  it('should getFilesByExtName works correctly', async () => {
    const files = await getFilesByExtName(srcDir, 'md');
    expect(files).to.be.an('array').that.includes('README.md');
  });

});


test/index.test.js
const path = require('path');
process.env.SHELL_NODE_MODULES_PATH = path.join(process.cwd(), 'node_modules');
process.env.CONFIGER_FOLDER_PATH = path.join(process.cwd(), 'node_modules');

const { expect } = require('chai');
const { getProdConfig, getDevConfig, DEFAULT_PROJECT_CONFIG } = require('../index');

describe('exports methods: <<getProdConfig>> and <<getDevConfig>>', () => {

  it('<<getProdConfig>> is a function', () => {
    expect(getProdConfig).to.be.a('function');
  });
  it('<<getDevConfig>> is a function', () => {
    expect(getDevConfig).to.be.a('function');
  });

  it('should <<getProdConfig>> return a promise', () => {
    expect(getProdConfig({})).to.be.a('promise');
  });
  it('should <<getDevConfig>> return a promise', () => {
    expect(getDevConfig({})).to.be.a('promise');
  });

  it('should getDevConfig a valid webpack config', async () => {
    const devConfig = await getDevConfig(DEFAULT_PROJECT_CONFIG, {
      port: 3000,
    });
    expect(devConfig).to.have.all.keys(
      'entry',
      'output',
      'resolve',
      'module',
      'externals',
      'target',
      'plugins',
      'node',
      'performance',
      'devtool',
    );
    expect(devConfig.resolve.extensions).to.eql(
      ['.web.js', '.js', '.json', '.jsx']
    );
    expect(devConfig.module.strictExportPresence).to.be.true;
    expect(devConfig.module.rules[0].oneOf).to.have.lengthOf(6);
    expect(devConfig.plugins).to.have.lengthOf(6);
  });

  it('should getProdConfig a valid webpack config', async () => {
    const prodConfig = await getProdConfig(DEFAULT_PROJECT_CONFIG);
    expect(prodConfig).to.have.all.keys(
      'bail',
      'entry',
      'output',
      'resolve',
      'module',
      'externals',
      'target',
      'plugins',
      'node',
      'devtool',
    );
    expect(prodConfig.resolve.extensions).to.eql(
      ['.web.js', '.js', '.json', '.jsx']
    );
    expect(prodConfig.module.strictExportPresence).to.be.true;
    expect(prodConfig.module.rules[0].oneOf).to.have.lengthOf(6);
    expect(prodConfig.plugins).to.have.lengthOf(8);
  });

});

index.js
/**
 * merge and export
 * @author ryan.bian
 */
const merge = require('webpack-merge');

const { PROJECT_ROOT } = require('./src/lib');
const _getBaseConfig = require('./src/base.conf');
const _getProdConfig = require('./src/prod.conf');
const _getDevConfig = require('./src/dev.conf');

const DEFAULT_PROJECT_CONFIG = require('./src/projectConfig');
const MERGE_STRATEGY = {
  plugins: 'append',
};

exports.DEFAULT_PROJECT_CONFIG = DEFAULT_PROJECT_CONFIG;

exports.getProdConfig = async (PROJECT_CONFIG, options) => {
  try {
    const baseConfig = await _getBaseConfig(PROJECT_CONFIG, options);
    const prodConfig = await _getProdConfig(PROJECT_CONFIG, options);
    return merge.smartStrategy(MERGE_STRATEGY)(baseConfig, prodConfig);
  } catch(e) {
    throw Error(e);
  }
};

exports.getDevConfig = async (PROJECT_CONFIG, options) => {
  try {
    const baseConfig = await _getBaseConfig(PROJECT_CONFIG, options);
    const devConfig = await _getDevConfig(PROJECT_CONFIG, options);
    return merge.smartStrategy(MERGE_STRATEGY)(baseConfig, devConfig);
  } catch(e) {
    throw Error(e);
  }
};

src/lib.js
const path = require('path');
const fs = require('fs');
const { promisify } = require('util');
const SHELL_NODE_MODULES_PATH = process.env.SHELL_NODE_MODULES_PATH;
const webpack = require(path.join(SHELL_NODE_MODULES_PATH, 'webpack'));

const PROJECT_ROOT = exports.PROJECT_ROOT = process.cwd();
const SOURCE_DIR = exports.SOURCE_DIR = path.join(PROJECT_ROOT, 'src');
const APP_DIR = exports.APP_DIR = path.join(SOURCE_DIR, 'app');

const stat = promisify(fs.stat);
const readdir = promisify(fs.readdir);

exports.getFilesByExtName = async (pathname, extname) => {
  const extnames = Array.isArray(extname) ? extname : [ extname ];
  const fileList = [];
  try {
    const files = await readdir(pathname);
    for (const file of files) {
      const filepath = path.join(pathname, file);
      if (extnames.some(ext => file.endsWith(`.${ext}`))) {
        const fileStat = await stat(filepath);
        if (fileStat.isFile()) {
          fileList.push(file);
        }
      }
    }
  } catch(e) {}

  return fileList;
};

exports.getHtmlLoaderConfig = PROJECT_CONFIG => ({
  test: /\.html?$/,
  use: [
    {
      loader: require.resolve('html-loader'),
      options: {
        interpolate: true,
        root: './',
      },
    },
    {
      loader: require.resolve('posthtml-loader'),
    },
  ],
});

exports.getLoaderOptionPlugin = PROJECT_CONFIG => new webpack.LoaderOptionsPlugin({
  options: {
    posthtml(ctx) {
      return {
        plugins: [
          require('posthtml-expressions')({
            locals:{
              env: process.env.NODE_ENV, 
            }
          })
        ],
      };
    }
  },
});


src/base.conf.js
/**
 * Base Config
 * TODO:
 * 1. introduce eslint
 */
const path = require('path');
const SHELL_NODE_MODULES_PATH = process.env.SHELL_NODE_MODULES_PATH;
const CONFIGER_FOLDER_PATH = process.env.CONFIGER_FOLDER_PATH;
const webpack = require(path.join(SHELL_NODE_MODULES_PATH, 'webpack'));
const { SOURCE_DIR } = require('./lib');

const CaseSensitivePathsPlugin = require('case-sensitive-paths-webpack-plugin');

module.exports = async (PROJECT_CONFIG, options) => {
  return {
    entry: {},
    output: {},
    resolve: {
      // These are the reasonable defaults supported by the Node ecosystem.
      // We also include JSX as a common component filename extension to support
      // some tools, although we do not recommend using it, see:
      // https://github.com/facebookincubator/create-react-app/issues/290
      // `web` extension prefixes have been added for better support
      // for React Native Web.
      extensions: ['.web.js', '.js', '.json', '.jsx'],
      modules: [
        'node_modules',
        path.join(CONFIGER_FOLDER_PATH, 'node_modules'),
      ],
    },
    module: {
      strictExportPresence: true,
    },
    externals: {},
    target: 'web',
    plugins: [
      new webpack.DefinePlugin({
        'process.env.NODE_ENV': JSON.stringify(process.env.NODE_ENV),
        'process.env.DEBUG': JSON.stringify(process.env.DEBUG)
      }),
      new CaseSensitivePathsPlugin(),
      new webpack.IgnorePlugin(/^\.\/locale$/, /moment$/),
    ],
    node: {
      dgram: 'empty',
      fs: 'empty',
      net: 'empty',
      tls: 'empty',
      child_process: 'empty',
    },
  };
};


src/prod.conf.js
/**
 * production config
 */
const path = require('path');
const SHELL_NODE_MODULES_PATH = process.env.SHELL_NODE_MODULES_PATH;
const webpack = require(path.join(SHELL_NODE_MODULES_PATH, 'webpack'));
const HtmlWebpackPlugin = require(path.join(SHELL_NODE_MODULES_PATH, 'html-webpack-plugin'));
const ExtractTextPlugin = require(path.join(SHELL_NODE_MODULES_PATH, 'extract-text-webpack-plugin'));
const CleanWebpackPlugin = require(path.join(SHELL_NODE_MODULES_PATH, 'clean-webpack-plugin'));
const { camelCase } = require('lodash');
const autoprefixer = require('autoprefixer');
const WebpackChunkHash = require('webpack-chunk-hash');

const {
  PROJECT_ROOT,
  APP_DIR,
  SOURCE_DIR,
  getFilesByExtName,
  getHtmlLoaderConfig,
  getLoaderOptionPlugin,
} = require('./lib');
const PUBLIC_PATH = '/';

module.exports = async (PROJECT_CONFIG, options) => {
  const BUILD_PATH = path.join(PROJECT_ROOT, PROJECT_CONFIG.buildPath);
  
  const configResult = {};
  
  const fileList = await getFilesByExtName(APP_DIR, ['html', 'htm']);
  const appPages = fileList.map(page => ({
      page,
      name: camelCase(page.replace(/\.html?$/, '')),
    }));

  // entry config
  const entry = {};

  appPages.forEach(d => {
    const scripts = [
      path.join(APP_DIR, `${d.name}.js`),
    ];
    Object.assign(entry, {
      [d.name]: scripts,
    });
  });

  // output config
  const output = {
    path: BUILD_PATH,
    filename: '[name].[chunkhash:8].js',
    publicPath: PROJECT_CONFIG.publicPath || PUBLIC_PATH,
  };

  const babelLoaderConfig = {
    loader: require.resolve('babel-loader'),
    options: {
      // @remove-on-eject-begin
      babelrc: false,
      presets: [
        [
          require.resolve('babel-preset-env'),
          {
            targets: {
              browsers: PROJECT_CONFIG.browserSupports.PRODUCTION,
            }, 
            module: false,
            useBuiltIns: PROJECT_CONFIG.babelUseBuiltIns,
          }
        ]
      ].concat(
        PROJECT_CONFIG.framework === 'react' ? [
          require.resolve('babel-preset-react'),
          require.resolve('babel-preset-stage-1'),
        ] : [
          require.resolve('babel-preset-stage-1'),
        ]
      ),
      // @remove-on-eject-end
      compact: true,
    },
  };

  // module config
  const module = {
    rules: [
      {
        // "oneOf" will traverse all following loaders until one will
        // match the requirements. When no loader matches it will fall
        // back to the "file" loader at the end of the loader list.
        oneOf: [
          // "url" loader works like "file" loader except that it embeds assets
          // smaller than specified limit in bytes as data URLs to avoid requests.
          // A missing `test` is equivalent to a match.
          {
            test: [/\.bmp$/, /\.gif$/, /\.jpe?g$/, /\.png$/],
            loader: require.resolve('url-loader'),
            options: {
              limit: 10000,
              name: '[name].[hash:8].[ext]',
            },
          },
          {
            test: /\.svg$/,
            include: SOURCE_DIR,
            resourceQuery: /reactComponent/,
            use: [
              babelLoaderConfig,
              {
                loader: require.resolve('react-svg-loader'),
                options: {
                  svgo: {
                    floatPrecision: 2,
                    plugins: [{
                      cleanupIDs: false,
                    }],
                  },
                },
              },
            ],
          },
          // Process JS with Babel.
          Object.assign({
            test: /\.jsx?$/,
            include: SOURCE_DIR,
          }, babelLoaderConfig),
          {
            test: /\.(le|c)ss$/,
            loader: ExtractTextPlugin.extract({
              fallback: {
                loader: require.resolve('style-loader'),
                options: {
                  hmr: false,
                },
              },
              use: [
                {
                  loader: require.resolve('css-loader'),
                  options: {
                    importLoaders: 1,
                    minimize: true,
                  },
                },
                {
                  loader: require.resolve('postcss-loader'),
                  options: {
                    // Necessary for external CSS imports to work
                    // https://github.com/facebookincubator/create-react-app/issues/2677
                    ident: 'postcss',
                    plugins: () => [
                      autoprefixer({
                        browsers: PROJECT_CONFIG.browserSupports.PRODUCTION,
                      }),
                    ],
                  },
                },
                {
                  loader: require.resolve('less-loader'),
                }
              ],
            }),
          },
          getHtmlLoaderConfig(PROJECT_CONFIG),
          // "file" loader makes sure assets end up in the `build` folder.
          // When you `import` an asset, you get its filename.
          // This loader doesn't use a "test" so it will catch all modules
          // that fall through the other loaders.
          {
            exclude: [/\.jsx?$/, /\.json$/],
            loader: require.resolve('file-loader'),
            options: {
              name: '[name].[hash:8].[ext]',
            },
          },
        ],
      }
    ],
  };

  // plugins config
  const plugins = [];
  
  appPages.forEach(d => {
    plugins.push(
      new HtmlWebpackPlugin(Object.assign({
        filename: d.page,
        template: path.join(APP_DIR, d.page),
        chunks: [
          d.name,
        ],
        minify: {
          removeComments: true,
          collapseWhitespace: true,
          removeRedundantAttributes: true,
          useShortDoctype: true,
          removeEmptyAttributes: false,
          removeStyleLinkTypeAttributes: true,
          keepClosingSlash: true,
          minifyJS: true,
          minifyCSS: true,
          minifyURLs: true,
        },
      }, PROJECT_CONFIG.htmlWebpackPlugin))
    );
  });
  plugins.push(
    new CleanWebpackPlugin([
      PROJECT_CONFIG.buildPath,
    ], {
      root: PROJECT_ROOT,
      verbose: true,
      dry: false,
    }),
    new webpack.HashedModuleIdsPlugin(),
    new WebpackChunkHash(),
    getLoaderOptionPlugin(PROJECT_CONFIG),
    new ExtractTextPlugin({
      filename: '[name].[contenthash:8].css',
    }),
  );

  Object.assign(configResult, {
    // Don't attempt to continue if there are any errors.
    bail: true,
    entry,
    output,
    module,
    plugins,
    devtool: 'source-map',
  });

  return configResult;
};

src/projectConfig.js
/**
 * Default Project Config
 */
module.exports = {
  type: 'spa',
  buildPath: 'dist',
  enableHotModuleReplacement: true,
  framework: 'react',
  htmlWebpackPlugin: {
    // true | 'head' | 'body' | false Inject all assets into the given template or templateContent
    // When passing true or 'body' all javascript resources will be placed at the bottom of the body element.
    // 'head' will place the scripts in the head element.
    inject: true,
    // Allows to control how chunks should be sorted before they are included to the html.
    // Allowed values: 'none' | 'auto' | 'dependency' |'manual' | {function}
    chunksSortMode: 'auto',
    // if true (default) try to emit the file only if it was changed.
    cache: true,
    // if true (default) errors details will be written into the HTML page.
    showErrors: true,
  },
  browserSupports: {
    DEVELOPMENT: ['last 2 versions'],
    PRODUCTION: ['last 2 versions'],
  },
  dll: {
    enable: false,
    include: [],
  },
  providePluginConfig: {
  },
  babelUseBuiltIns: true,
  // valid in dev environment
  https: false,
  publicPath: '../',
};


src/dev.conf.js
/**
 * development config
 */
const path = require('path');
const SHELL_NODE_MODULES_PATH = process.env.SHELL_NODE_MODULES_PATH;
const webpack = require(path.join(SHELL_NODE_MODULES_PATH, 'webpack'));
const HtmlWebpackPlugin = require(path.join(SHELL_NODE_MODULES_PATH, 'html-webpack-plugin'));
const { camelCase } = require('lodash');
const autoprefixer = require('autoprefixer');

const {
  PROJECT_ROOT,
  APP_DIR,
  getFilesByExtName,
  SOURCE_DIR,
  getHtmlLoaderConfig,
  getLoaderOptionPlugin,
} = require('./lib');
const PUBLIC_PATH = '/';

module.exports = async (PROJECT_CONFIG, options) => {
  const BUILD_PATH = path.join(PROJECT_ROOT, PROJECT_CONFIG.buildPath);
  
  const configResult = {};
  
  const fileList = await getFilesByExtName(APP_DIR, ['html', 'htm']);
  const appPages = fileList.map(page => ({
    page,
    name: camelCase(page.replace(/\.html?$/, '')),
  }));

  // entry config
  const entry = {};
  const devServerEntry = [
    `${require.resolve(`${path.join(SHELL_NODE_MODULES_PATH, 'webpack-dev-server')}/client`)}?http://localhost:${options.port}`,
    require.resolve(`${path.join(SHELL_NODE_MODULES_PATH, 'webpack')}/hot/dev-server`),
  ];
  if (PROJECT_CONFIG.framework === 'react') {
    devServerEntry.unshift(require.resolve('react-hot-loader/patch'));
  }

  appPages.forEach(d => {
    const scripts = [
      path.join(APP_DIR, `${d.name}.js`),
    ];
    if (PROJECT_CONFIG.enableHotModuleReplacement) {
      scripts.unshift(...devServerEntry);
    }
    Object.assign(entry, {
      [d.name]: scripts,
    });
  });

  // output config
  const output = {
    path: BUILD_PATH,
    // Add /* filename */ comments to generated require()s in the output.
    pathinfo: true,
    publicPath: PUBLIC_PATH,
  };

  const babelLoaderConfig = {
    loader: require.resolve('babel-loader'),
    options: {
      // @remove-on-eject-begin
      babelrc: false,
      presets: [
        [
          require.resolve('babel-preset-env'),
          {
            targets: {
              browsers: PROJECT_CONFIG.browserSupports.DEVELOPMENT,
            }, 
            module: false,
            useBuiltIns: PROJECT_CONFIG.babelUseBuiltIns,
          }
        ]
      ].concat(
        PROJECT_CONFIG.framework === 'react' ? [
          require.resolve('babel-preset-react'),
          require.resolve('babel-preset-stage-1'),
        ] : [
          require.resolve('babel-preset-stage-1'),
        ]
      ),
      // @remove-on-eject-end
      // This is a feature of `babel-loader` for webpack (not Babel itself).
      // It enables caching results in ./node_modules/.cache/babel-loader/
      // directory for faster rebuilds.
      cacheDirectory: true,
    },
  }

  // module config
  const module = {
    rules: [
      {
        // "oneOf" will traverse all following loaders until one will
        // match the requirements. When no loader matches it will fall
        // back to the "file" loader at the end of the loader list.
        oneOf: [
          // "url" loader works like "file" loader except that it embeds assets
          // smaller than specified limit in bytes as data URLs to avoid requests.
          // A missing `test` is equivalent to a match.
          {
            test: [/\.bmp$/, /\.gif$/, /\.jpe?g$/, /\.png$/],
            loader: require.resolve('url-loader'),
            options: {
              limit: 10000,
              name: '[name].[hash:8].[ext]',
            },
          },
          {
            test: /\.svg$/,
            include: SOURCE_DIR,
            resourceQuery: /reactComponent/,
            use: [
              babelLoaderConfig,
              {
                loader: require.resolve('react-svg-loader'),
                options: {
                  svgo: {
                    floatPrecision: 2,
                    plugins: [{
                      cleanupIDs: false,
                    }],
                  },
                },
              },
            ],
          },
          // Process JS with Babel.
          Object.assign({
            test: /\.jsx?$/,
            include: SOURCE_DIR,
          }, babelLoaderConfig),
          {
            test: /\.(le|c)ss$/,
            use: [
              require.resolve('style-loader'),
              {
                loader: require.resolve('css-loader'),
                options: {
                  importLoaders: 1,
                  minimize: false,
                },
              },
              {
                loader: require.resolve('postcss-loader'),
                options: {
                  // Necessary for external CSS imports to work
                  // https://github.com/facebookincubator/create-react-app/issues/2677
                  ident: 'postcss',
                  plugins: () => [
                    autoprefixer({
                      browsers: PROJECT_CONFIG.browserSupports.DEVELOPMENT,
                    }),
                  ],
                },
              },
              {
                loader: require.resolve('less-loader'),
              }
            ],
          },
          getHtmlLoaderConfig(PROJECT_CONFIG),
          // "file" loader makes sure those assets get served by WebpackDevServer.
          // When you `import` an asset, you get its (virtual) filename.
          // In production, they would get copied to the `build` folder.
          // This loader doesn't use a "test" so it will catch all modules
          // that fall through the other loaders.
          {
            // Exclude `js` files to keep "css" loader working as it injects
            // it's runtime that would otherwise processed through "file" loader.
            // Also exclude `html` and `json` extensions so they get processed
            // by webpacks internal loaders.
            exclude: [/\.jsx?$/, /\.json$/],
            loader: require.resolve('file-loader'),
            options: {
              name: '[name].[hash:8].[ext]',
            },
          },
        ],
      }
    ]
  };

  // plugins config
  const plugins = [];
  
  appPages.forEach(d => {
    plugins.push(
      new HtmlWebpackPlugin(Object.assign({
        filename: d.page,
        template: path.join(APP_DIR, d.page),
        chunks: [
          d.name,
        ],
        minify: {
          removeComments: true,
          collapseWhitespace: true,
          removeRedundantAttributes: true,
          useShortDoctype: true,
          removeEmptyAttributes: false,
          removeStyleLinkTypeAttributes: true,
          keepClosingSlash: true,
          minifyJS: true,
          minifyCSS: true,
          minifyURLs: true,
        },
      }, PROJECT_CONFIG.htmlWebpackPlugin))
    );
  });
  plugins.push(
    // Add module names to factory functions so they appear in browser profiler.
    new webpack.NamedModulesPlugin(),
    getLoaderOptionPlugin(PROJECT_CONFIG),
  );
  if (PROJECT_CONFIG.enableHotModuleReplacement) {
    plugins.push(
      // This is necessary to emit hot updates (currently CSS only):
      new webpack.HotModuleReplacementPlugin(),
    );
  }

  Object.assign(configResult, {
    entry,
    output,
    module,
    plugins,
    devtool: 'cheap-module-source-map',
    // Turn off performance hints during development because we don't do any
    // splitting or minification in interest of speed. These warnings become
    // cumbersome.
    performance: {
      hints: false,
    },
  });

  return configResult;
};